crates/iceberg/src/expr/visitors/manifest_evaluator.rs

```diff
@@ -103,98 +106,255 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
+            Some(bound) if datum > bound => ROWS_CANNOT_MATCH,
+            Some(_) => ROWS_MIGHT_MATCH,
+            None => ROWS_CANNOT_MATCH,
+        }
     }
 
     fn eq(
         &mut self,
         reference: &BoundReference,
-        literal: &Datum,
+        datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field = self.field_summary_for_reference(reference);
+
+        if field.lower_bound.is_none() || field.upper_bound.is_none() {
+            return ROWS_CANNOT_MATCH;
+        }
+
+        if let Some(lower_bound) = &field.lower_bound {
+            if lower_bound > datum {
+                return ROWS_CANNOT_MATCH;
+            }
+        }
+
+        if let Some(upper_bound) = &field.upper_bound {
+            if upper_bound < datum {
+                return ROWS_CANNOT_MATCH;
+            }
+        }
+
+        ROWS_MIGHT_MATCH
     }
 
     fn not_eq(
         &mut self,
-        reference: &BoundReference,
-        literal: &Datum,
+        _reference: &BoundReference,
+        _datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        // because the bounds are not necessarily a min or max value, this cannot be answered using
+        // them. notEq(col, X) with (X, Y) doesn't guarantee that X is a value in col.
+        ROWS_MIGHT_MATCH
     }
 
     fn starts_with(
         &mut self,
         reference: &BoundReference,
-        literal: &Datum,
+        datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field = self.field_summary_for_reference(reference);
+
+        if field.lower_bound.is_none() || field.upper_bound.is_none() {
+            return ROWS_CANNOT_MATCH;
+        }
+
+        let PrimitiveLiteral::String(prefix) = datum.literal() else {
+            return Err(Error::new(
+                ErrorKind::Unexpected,
+                "Cannot perform starts_with on non-string value",
+            ));
+        };
+
+        let prefix_len = prefix.len();
+
+        if let Some(lower_bound) = &field.lower_bound {
+            let PrimitiveLiteral::String(lower_bound) = lower_bound.literal() else {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    "Cannot perform starts_with on non-string lower bound",
+                ));
+            };
+
+            let min_len = lower_bound.len().min(prefix_len);
+            if prefix.as_bytes().lt(&lower_bound.as_bytes()[..min_len]) {
+                return ROWS_CANNOT_MATCH;
+            }
+        }
+
+        if let Some(upper_bound) = &field.upper_bound {
+            let PrimitiveLiteral::String(upper_bound) = upper_bound.literal() else {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    "Cannot perform starts_with on non-string upper bound",
+                ));
+            };
```

Comment:
@marvinlanhenke, I looked at reusing some of the code, but because we have custom checks, I could only extract this portion of the code (e.g. lines 283 to 289) for lower bound and upper bound in separate utility functions. Once I do that then, I still have to unwrap the returned value if it wasn't an error. The new code doesn't make the existing code much cleaner, in my opinion. But if you think it would add value then, I can push the new changes.

Reply:
> (e.g. lines 283 to 289)

@s-akhtar-baig those were the lines I was referring to. 

Something like that might work?

```Rust
// mostly pseudo code
fn check_boundaries(field: &FieldSummary) -> Result<&String> {
  match &field.lower_bound {
    Some(lower_bound) => {
      let PrimitiveLiteral::String(lower_bound) = lower_bound.literal() else {
        return Err(Error::new(ErrorKind::Unexpected, "Cannot perform..."));
      }
      Ok(lower_bound)
    }
   None => Err(Error::new(...))
  }
}
```

If that works you can probably create a small macro to avoid having a lower and upper bound variant?
```Rust
macro_rules! check_boundaries {
    ($field:expr, $boundary:ident) => {
        match &$field.$boundary {
            Some(bound_value) => {
//...the rest
```

Reply:
> (e.g. lines 283 to 289)

@s-akhtar-baig those were the lines I was referring to. 
Perhaps something like this might work?

```Rust
// this should be enough?
fn check_bounds(datum: &Datum ) -> Result<&String> {
  let PrimitiveLiteral::String(boundary) = datum.literal() else {
    return Err(Error::new(ErrorKind::Unexpected, "Cannot perform..."));
  };
  Ok(boundary)
}

// and then
if let Some(upper_bound) = &field.upper_bound {
  let boundary = Self::check_bounds(upper_bound)?;
  // ...proceed with custom checks
}
```

Reply:
Thanks @marvinlanhenke! Please let me know if the latest changes reflect your suggestions correctly.

Reply:
> Thanks @marvinlanhenke! Please let me know if the latest changes reflect your suggestions correctly.

Yes this looks great 👌 thanks for your effort.